File: go/aoc-day-10/aoc-day-10.go
package main

import (
	"bufio"   // Scanner for reading file
	"fmt"     // Println, Printf, Sprintf, etc.
	"os"      // File operations
	"strconv" // Atoi, Itoa, etc.
)

type Tuple struct {
	x int
	y int
}

func main() {
	file, err := os.Open("topo-map.txt")
	if err != nil {
		fmt.Println("Error reading file: ", err)
		return
	}
	// Close the file when the function returns
	defer file.Close()

	// slice of byte slices
	matrix := [][]byte{}
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		// text line from the file
		text := scanner.Text()

		// convert the text line to a byte array
		row := []byte(text)

		// append the byte array to the matrix
		matrix = append(matrix, row)
	}

	// Part 1
	part_1(matrix)

	// Part 2
	part_2(matrix)
}

func part_1(matrix [][]byte) {
	trail_count := 0
	visited := make(map[Tuple]bool)
	for i := 0; i < len(matrix); i++ {
		for j := 0; j < len(matrix[i]); j++ {
			// conversion from byte to char is implicit
			if matrix[i][j] == '0' {

				// clear visited map
				visited = make(map[Tuple]bool)
				trail_count += dfs_with_set(matrix, i, j, 0, visited)
			}
		}
	}
	fmt.Println("Number of trails (Part 1): ", trail_count)
}

func part_2(matrix [][]byte) {
	trail_count := 0
	for i := 0; i < len(matrix); i++ {
		for j := 0; j < len(matrix[i]); j++ {
			// conversion from byte to char is implicit
			if matrix[i][j] == '0' {
				trail_count += dfs(matrix, i, j, 0)
			}
		}
	}
	fmt.Println("Number of trails (Part 2): ", trail_count)
}

func dfs_with_set(matrix [][]byte, i int, j int, index int, visited map[Tuple]bool) int {
	if i < 0 || j < 0 || i >= len(matrix) || j >= len(matrix[i]) || matrix[i][j] != strconv.Itoa(index)[0] {
		return 0
	}

	_, ok := visited[Tuple{i, j}]
	if matrix[i][j] == '9' && !ok {
		visited[Tuple{i, j}] = true
		return 1
	}

	count := 0
	count += dfs_with_set(matrix, i+1, j, index+1, visited)
	count += dfs_with_set(matrix, i-1, j, index+1, visited)
	count += dfs_with_set(matrix, i, j+1, index+1, visited)
	count += dfs_with_set(matrix, i, j-1, index+1, visited)

	return count
}

func dfs(matrix [][]byte, i int, j int, index int) int {
	if i < 0 || j < 0 || i >= len(matrix) || j >= len(matrix[i]) || matrix[i][j] != strconv.Itoa(index)[0] {
		return 0
	}

	if matrix[i][j] == '9' {
		return 1
	}

	count := 0
	count += dfs(matrix, i+1, j, index+1)
	count += dfs(matrix, i-1, j, index+1)
	count += dfs(matrix, i, j+1, index+1)
	count += dfs(matrix, i, j-1, index+1)

	return count
}

File: python/aoc-day-10/aoc-day-10.py
# Bad practice but it does the job
visited = set()


def main():
    with open("topo-map.txt", "r") as file:
        topo_map = [list(line.strip()) for line in file]
    # print(topo_map)

    # Part 1
    part_1(topo_map)

    # Part 2
    part_2(topo_map)


def part_1(topo_map):
    possible_trails = 0
    for i in range(len(topo_map)):
        for j in range(len(topo_map[i])):
            if topo_map[i][j] == "0":
                # Reset the visited 9s set
                visited.clear()
                dfs_map = topo_map[:]
                trails = dfs_with_set(topo_map, i, j, 0)
                possible_trails += trails

    print(possible_trails)


def part_2(topo_map):
    possible_trails = 0
    for i in range(len(topo_map)):
        for j in range(len(topo_map[i])):
            if topo_map[i][j] == "0":
                dfs_map = topo_map[:]
                trails = dfs(topo_map, i, j, 0)
                possible_trails += trails

    print(possible_trails)


def dfs(topo_map, i, j, index):
    if i < 0 or j < 0 or i >= len(topo_map) or j >= len(topo_map[i]):
        return 0
    if topo_map[i][j] != str(index):
        return 0
    elif topo_map[i][j] == "9":
        return 1
    else:
        # Same reasoning as in dfs_with_set
        topo_map[i][j] = "X"
        possible_trails = 0

        # Check all possible directions if they are in bounds for next number in sequence (index + 1)
        possible_trails += dfs(topo_map, i - 1, j, index + 1)
        possible_trails += dfs(topo_map, i + 1, j, index + 1)
        possible_trails += dfs(topo_map, i, j - 1, index + 1)
        possible_trails += dfs(topo_map, i, j + 1, index + 1)

        # Unvisit the node
        topo_map[i][j] = str(index)
        return possible_trails


def dfs_with_set(topo_map, i, j, index):
    if i < 0 or j < 0 or i >= len(topo_map) or j >= len(topo_map[i]):
        return 0
    if topo_map[i][j] != str(index):
        return 0
    elif topo_map[i][j] == "9" and (i, j) not in visited:
        # Visit the node, no more trail from this distince 0 node can lead to this distinct 9 node
        visited.add((i, j))
        return 1
    else:
        # Mark the node as visited, not this is not the same as the visited set
        # Since we are using recursive dfs, we need to mark for sake of not
        # having a stack overflow, visited set makes sure we don't count a 9 node twice
        topo_map[i][j] = "X"
        possible_trails = 0

        # Check all possible directions if they are in bounds for next number in sequence (index + 1)
        possible_trails += dfs_with_set(topo_map, i - 1, j, index + 1)
        possible_trails += dfs_with_set(topo_map, i + 1, j, index + 1)
        possible_trails += dfs_with_set(topo_map, i, j - 1, index + 1)
        possible_trails += dfs_with_set(topo_map, i, j + 1, index + 1)

        # Unvisit the node
        topo_map[i][j] = str(index)
        return possible_trails


if __name__ == "__main__":
    main()
